### PR TITLE
Add unambiguous discrimination strategy to state distinguishability and state exclusion

### LINKED ISSUE #519: Add unambiguous discrimination strategy to state distinguishability and state exclusion
Enable the ability for the user to unambiguously exclude or discriminate amongst of quantum states in the `state_distinguishability.py` and `state_exclusion.py` files, respectively. 

Refer to Section 2.4 in [arXiv:1707.02571](https://arxiv.org/pdf/1707.02571.pdf) for further information on the constraints and definition of the unambiguous case for the SDP and refer to `ppt_distinguishability.py` for an example of how the unambiguous strategy is supported within the SDP and function arguments.

### PR DESCRIPTION
**Pull request by tnemoz**
## Description
Enable the ability for the user to unambiguously exclude or discriminate amongst of quantum states in the state_distinguishability.py and state_exclusion.py files, respectively. Fixes #519 

## Changes
Notable changes that this PR has either accomplished or will accomplish. Feel free to add more lines to the itemized list
below.

  - [x] More tests have been added to the tests suite 
  - [x] The user now has the possibility to unambiguously distinguish/exclude states via the implementation of the corresponding SDPs
  - [x] The user now has the possibility to pass arbitrary options to picos' solver. This fixes current, unreported bugs in `state_exclusion.py`

## Checklist
Before marking your PR ready for review, make sure you checked the following locally. If this is your first PR, you might be notified of some workflow failures after a maintainer has approved the workflow jobs to be run on your PR. 

Additional information is available in the [documentation](https://toqito.readthedocs.io/en/latest/contributing.html#testing).

  -  [x] Use `ruff` for errors related to code style and formatting.
  -  [x] Verify all previous and newly added unit tests pass in `pytest`.
  -  [x] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR
  -  [x] Use `linkcheck` to check for broken links in the documentation
  -  [x] Use `doctest` to verify the examples in the function docstrings work as expected.

### CODE REVIEW

File: toqito/state_opt/state_exclusion.py
Comment (purva-thakre) [inline toqito/state_opt/state_exclusion.py:142]: If you are adding a link as a reference, add it to `docs/articles.bib` and then cite it. Look for how we add Wikipedia links in this file.

File: toqito/state_opt/state_exclusion.py
Comment (purva-thakre) [inline toqito/state_opt/state_exclusion.py:241]: Add the `arxiv` reference to `articles.bib` and then cite it.

File: toqito/state_opt/state_distinguishability.py
Comment (purva-thakre) [inline toqito/state_opt/state_distinguishability.py:23]: Thanks for catching this. Instead of changing `rho_0` to `rho_1`, I think it might be easier to change `rho_n` to `rho_{n-1}`

File: toqito/state_opt/state_distinguishability.py
Comment (purva-thakre) [inline toqito/state_opt/state_distinguishability.py:28]: Same here as the `rho_0` <-> `rho_1` comment.

File: toqito/state_opt/state_distinguishability.py
Comment (vprusso) [reply toqito/state_opt/state_distinguishability.py:23]: I think it might actually be better to do 1 -> n. Part of the reason for this is this would require making changes elsewhere in the docstring that I think would be cumbersome/confusing. 

I think this numbering scheme is distinct from any confusion surrounding the Python indexing, so I'd personally prefer 1 -> n

Comment (purva-thakre) [reply toqito/state_opt/state_distinguishability.py:23]: Fine by me! I was trying to make it consistent with python indexing.

File: toqito/state_opt/state_distinguishability.py
Comment (purva-thakre) [reply toqito/state_opt/state_distinguishability.py:28]: Resolving due to https://github.com/vprusso/toqito/pull/614#discussion_r1624748864

File: toqito/state_opt/state_exclusion.py
Comment (vprusso) [reply toqito/state_opt/state_exclusion.py:241]: Yeah, this I agree that we should cite from the bib (just not for the GitHub/GitLab issue).

File: toqito/state_opt/state_exclusion.py
Comment (purva-thakre) [reply toqito/state_opt/state_exclusion.py:142]: That's how I added one such issue: https://github.com/vprusso/toqito/blob/49ecfd489d8be0fea23f88291779a2ef03065d23/docs/articles.bib#L217

Comment (vprusso) [reply toqito/state_opt/state_exclusion.py:142]: Gotcha. My preference would be to keep these types of things in line, as the issues are somewhat ephemeral once they are closed out. 

As an aside, (and this is likely a separate issue that I should put on the board) but I think it makes sense to consolidate the references into a single `.bib` file. Not that `mitiq` needs to be the gold standard, but in seeing how they arrange things, it seems like a single `refs.bib` is used and I don't see any gitlab/github URLs located in those reference files. 

I can see an argument for keeping the URL in the bib file, but my vote would be to keep them out.

**Comment by purva-thakre** [comment]
I don't have access to a `macOS` device either. I rely on using the ones available as Github runners. 

If you want to test some changes out without waiting for us to approve and run the workflow in this PR, you should be able to trigger the workflows in your fork of this repo. 

>  Do you know if there are some specificities in these versions that could explain the discrepancy?

Right now, I don't know what could be causing this. I can search around. We have noticed similar failures elsewhere https://github.com/vprusso/toqito/issues/613

**Comment by purva-thakre** [comment]
We could also temporarily skip the `macOS` tests to at least find the issue with python3.12 on linux. 

I have used the following in the past:
https://github.com/vprusso/toqito/blob/73641d84f25ab04e31fdf657ee97c64726d17b33/toqito/perms/tests/test_antisymmetric_projection.py#L20-L22

**Comment by tnemoz** [comment]
Hey @purva-thakre and @vprusso! The tests should be passing now (or at least they did on my own repo :smile:).

I must admit I'm quite unsatisfied by this though. CVXOPT will be the go-to option for most users and they currently have to sacrifice a lamb on a full moon to get the parameters right (OK, they just have to reduce the accuracy and/or change the solver, but that's still not really convenient). But I don't see a good solution to this problem for now :cry:

**Comment by vprusso** [comment]
> Hey @purva-thakre and @vprusso! The tests should be passing now (or at least they did on my own repo 😄).
> 
> I must admit I'm quite unsatisfied by this though. CVXOPT will be the go-to option for most users and they currently have to sacrifice a lamb on a full moon to get the parameters right (OK, they just have to reduce the accuracy and/or change the solver, but that's still not really convenient). But I don't see a good solution to this problem for now 😢

You mean you *don't* need to do the full moon sacrifice thing to get CVXOPT!? I've got to make some phone calls apparently...

Anyway, I definitely agree with your comment. It's absolutely less than ideal to need to reduce the accuracy as a solution. The MOSEK solver in practice is quite a bit more pleasant to use--it's much faster, more accurate, etc. But, as a downside, it's also not free :(

And yes, seems like the tests pass, which is great. There's some codecov thing, but that's something I need to worry about. I'll try and get to reviewing the code in a bit, and thank you again for your work thus far! Appreciate all the lamb sacrifices you've had to make to get here!

**Review by vprusso — APPROVED** [review]
Minor comments, but LGTM! I'll let @purva-thakre give the final say, but good on my end. Thank you again, @tnemoz for your work on this one!

**Review by purva-thakre — APPROVED** [review]
LGTM too! Once Vincent's suggestions are resolved, feel free to merge.

File: toqito/state_opt/state_exclusion.py
Comment (tnemoz) [reply toqito/state_opt/state_exclusion.py:87]: Fundamentally, I'm not even sure it should be a warning: it just happens that we can always succeed with probability $1$ in that case and that we know a closed-form expression. This could be transparent to the user if we directly set the variables to what they should be.

For the dual it's easy, we can take $N=\sum\limits_ip_i\rho_i$ and $a_i=1$.

For the primal however, I'm not sure (there's probably a closed-form expression, but I don't know it!)

Comment (vprusso) [reply toqito/state_opt/state_exclusion.py:87]: True, it probably doesn't need to be a warning--it's more so a property of the fact that linearly dependent quantum states will always yield a value of 1. Perhaps removing the warning and putting this info in the docstring would be okay?

Aside from that though, I'm happy with the PR and can merge whenever you feel it to be ready, @tnemoz !

**Comment by tnemoz** [comment]
I think it's all good now @vprusso and @purva-thakre! However, both of you said that I should feel free to merge, but I'm not sure I have the rights to do so :smile:

**Comment by purva-thakre** [comment]
> but I'm not sure I have the rights to do so 😄

Ah. I did not know that. @vprusso will have to look into the repo settings for this. 

I will merge once all checks pass.

**Comment by purva-thakre** [comment]
Congrats on the bounty! @tnemoz

**Comment by tnemoz** [comment]
> Congrats on the bounty! @tnemoz

Thanks! If we except the CVXOPT thing, it was really fun working on this! :smile:

And also, UnitaryHack made me discover this package and I find it pretty cool! If I have some time, I'll maybe try to do some (disinterested this time) PR! Notably, I would like to extend the tests to Mosek and change some code from CVXOPT to te more generic Picos.

**Comment by vprusso** [comment]
> > Congrats on the bounty! @tnemoz
> 
> Thanks! If we except the CVXOPT thing, it was really fun working on this! 😄
> 
> And also, UnitaryHack made me discover this package and I find it pretty cool! If I have some time, I'll maybe try to do some (disinterested this time) PR! Notably, I would like to extend the tests to Mosek and change some code from CVXOPT to the more generic Picos.

Thank you again for your time on this, @tnemoz 

I want you to know that your contributions and input are most welcome. While sometimes frustrating to work with, I agree that the CVXOPT package is quite cool! It's also come a long way and has only relatively recently integrated support for working over complex numbers (and hence applications pertinent to quantum information). 

You are always welcome here at any time during or after UnitaryHack! Don't be a stranger, and thank you again for your contribution!

**Pull request merged**
